File: 2019/Day07/intcode.py
import re


class Computer:


    def __init__(self, start_position=0, opcodes=[]):
        self.instruction_pointer = start_position
        self.ram = dict(enumerate(opcodes))
        self.parametric_opcodes = ("01", "02", "05", "06", "07", "08")
        self.register = 0
        self.instructions = {
            "1": self.add, "2": self.multiply, "3": self.receive_input, "4": self.output, "99": self.halt,
            "5": self.jump_if_true, "6": self.jump_if_false, "7": self.less_than, "8": self.equals,
            "01": self.add, "02": self.multiply, "03": self.receive_input, "04": self.output,
            "05": self.jump_if_true, "06": self.jump_if_false, "07": self.less_than, "08": self.equals,
        }
    
    
    def load_program(self, opcodes):
        self.ram = dict(enumerate(opcodes))
        self.instruction_pointer = 0
    
    
    def view_memory(self, pos):
        return self.ram[pos]
    

    def parse_instruction(self):
        ret_value = 0
        instruction = self.ram[self.instruction_pointer]
        in1_mode = in2_mode = out_mode = "0"
        if instruction == "4":
            in1_mode = in2_mode = out_mode = "0"
            self.instructions[instruction](in1_mode, in2_mode, out_mode)
            ret_value = 4
        elif len(instruction) == 1:
            in1_mode = in2_mode = out_mode = "0"
            if instruction in self.instructions.keys():
                self.instructions[instruction](in1_mode, in2_mode, out_mode)
            else:
                ret_value = 1
        elif instruction[-2:] in self.parametric_opcodes:
            in1_mode = in2_mode = out_mode = "0"
            if len(instruction) > 2:
                in1_mode = instruction[-3]
                if len(instruction) > 3:
                    in2_mode = instruction[-4]
            self.instructions[instruction[-2:]](in1_mode, in2_mode, out_mode)
        elif instruction[-2:] == "03":
            self.receive_input(in1_mode, in2_mode, out_mode)
        elif instruction[-2:] == "04":
            out_mode = "0"
            if len(instruction) == 3:
                out_mode = instruction[-3]
            self.output(in1_mode, in2_mode, out_mode)
            ret_value = 4
        elif instruction[-2:] == "99":
            self.halt(in1_mode, in2_mode, out_mode)
            ret_value = 99
        else:
            ret_value = 1
        return ret_value


    def run(self):
        out = self.parse_instruction()
        while out == 0:
            out = self.parse_instruction()
        if out == 1:
            print("Invalid opcode!")
            return 1
        elif out == 4:
            return 4
        elif out == 99:
            # print("Execution completed!")
            return 99
    

    def output(self, in1_mode, in2_mode, out_mode):
        if out_mode == "0":
            out_pos = int(self.ram[self.instruction_pointer + 1])
        else:
            out_pos = self.instruction_pointer + 1
        self.register = self.ram[out_pos]
        # print(self.register)
        self.instruction_pointer += 2


    def add(self, in1_mode, in2_mode, out_mode):
        if in1_mode == "0":
            in1_pos = int(self.ram[self.instruction_pointer + 1])
        else:
            in1_pos = self.instruction_pointer + 1
        if in2_mode == "0":
            in2_pos = int(self.ram[self.instruction_pointer + 2])
        else:
            in2_pos = self.instruction_pointer + 2
        in1 = int(self.ram[in1_pos])
        in2 = int(self.ram[in2_pos])
        out_pos = int(self.ram[self.instruction_pointer + 3])
        self.ram[out_pos] = str(in1 + in2)
        self.instruction_pointer += 4


    def multiply(self, in1_mode, in2_mode, out_mode):
        if in1_mode == "0":
            in1_pos = int(self.ram[self.instruction_pointer + 1])
        else:
            in1_pos = self.instruction_pointer + 1
        if in2_mode == "0":
            in2_pos = int(self.ram[self.instruction_pointer + 2])
        else:
            in2_pos = self.instruction_pointer + 2
        in1 = int(self.ram[in1_pos])
        in2 = int(self.ram[in2_pos])
        out_pos = int(self.ram[self.instruction_pointer + 3])
        self.ram[out_pos] = str(in1 * in2)
        self.instruction_pointer += 4


    def receive_input(self, in1_mode, in2_mode, out_mode):
        if self.inputs:
            val = self.inputs.pop()
        else:
            val = input("Input an integer: ")
        out_pos = int(self.ram[self.instruction_pointer + 1])
        self.ram[out_pos] = val
        self.instruction_pointer += 2
    

    def jump_if_true(self, in1_mode, in2_mode, out_mode):
        if in1_mode == "0":
            in1_pos = int(self.ram[self.instruction_pointer + 1])
        else:
            in1_pos = self.instruction_pointer + 1
        if in2_mode == "0":
            in2_pos = int(self.ram[self.instruction_pointer + 2])
        else:
            in2_pos = self.instruction_pointer + 2
        in1 = int(self.ram[in1_pos])
        in2 = int(self.ram[in2_pos])
        if in1 != 0:
            self.instruction_pointer = in2
        else:
            self.instruction_pointer += 3
    

    def jump_if_false(self, in1_mode, in2_mode, out_mode):
        if in1_mode == "0":
            in1_pos = int(self.ram[self.instruction_pointer + 1])
        else:
            in1_pos = self.instruction_pointer + 1
        if in2_mode == "0":
            in2_pos = int(self.ram[self.instruction_pointer + 2])
        else:
            in2_pos = self.instruction_pointer + 2
        in1 = int(self.ram[in1_pos])
        in2 = int(self.ram[in2_pos])
        if in1 == 0:
            self.instruction_pointer = in2
        else:
            self.instruction_pointer += 3


    def less_than(self, in1_mode, in2_mode, out_mode):
        if in1_mode == "0":
            in1_pos = int(self.ram[self.instruction_pointer + 1])
        else:
            in1_pos = self.instruction_pointer + 1
        if in2_mode == "0":
            in2_pos = int(self.ram[self.instruction_pointer + 2])
        else:
            in2_pos = self.instruction_pointer + 2
        in1 = int(self.ram[in1_pos])
        in2 = int(self.ram[in2_pos])
        out_pos = int(self.ram[self.instruction_pointer + 3])
        if in1 < in2:
            self.ram[out_pos] = "1"
        else:
            self.ram[out_pos] = "0"
        self.instruction_pointer += 4


    def equals(self, in1_mode, in2_mode, out_mode):
        if in1_mode == "0":
            in1_pos = int(self.ram[self.instruction_pointer + 1])
        else:
            in1_pos = self.instruction_pointer + 1
        if in2_mode == "0":
            in2_pos = int(self.ram[self.instruction_pointer + 2])
        else:
            in2_pos = self.instruction_pointer + 2
        in1 = int(self.ram[in1_pos])
        in2 = int(self.ram[in2_pos])
        out_pos = int(self.ram[self.instruction_pointer + 3])
        if in1 == in2:
            self.ram[out_pos] = "1"
        else:
            self.ram[out_pos] = "0"
        self.instruction_pointer += 4


    def halt(self, in1_mode, in2_mode, out_mode):
        pass


def opcodes_from_string(text):
    opcodes = [m.strip() for m in text.split(",")]
    return opcodes


if __name__ == "__main__":
    opcode_str = "3,12,6,12,15,1,13,14,13,4,13,99,-1,0,1,9"
    my_computer = Computer()
    my_computer.load_program(opcodes_from_string(opcode_str))
    my_computer.run()

File: 2019/Day09/intcode.rs
use std::fs;
use std::collections::HashMap;

enum State {
    IDLE,
    RUN,
    WAIT,
    HALT,
}

struct Computer {
    ram: HashMap<usize, i64>,
    state: State,
    instruction_pointer: usize,
    relative_base: usize,
    inputs: Vec<i64>,
    outputs: Vec<i64>,
}

impl Computer {
    fn initialize(&mut self) {
        self.state = State::RUN;
    }

    fn load_program(&mut self, program: &HashMap<usize, i64>) {
        self.ram = program.iter().map(|(k, v)| (*k, *v)).collect();
        self.instruction_pointer = 0;
        self.relative_base = 0;
    }

    fn run(&mut self) {
        let mut instruction: i64 = *self.ram.entry(self.instruction_pointer).or_insert(0);
        let opcode: i64 = instruction % 100;
        instruction /= 100;
        let in1_pos: usize = match instruction % 10 {
            0 => *self.ram.entry(self.instruction_pointer + 1).or_insert(0) as usize,
            1 => self.instruction_pointer + 1,
            2 => (*self.ram.entry(self.instruction_pointer + 1).or_insert(0) + self.relative_base as i64) as usize,
            _ => {
                println!("Invalid input 1 mode!");
                0
            }
        };
        instruction /= 10;
        let in2_pos: usize = match instruction % 10 {
            0 => *self.ram.entry(self.instruction_pointer + 2).or_insert(0) as usize,
            1 => self.instruction_pointer + 2,
            2 => (*self.ram.entry(self.instruction_pointer + 2).or_insert(0) + self.relative_base as i64) as usize,
            _ => {
                println!("Invalid input 2 mode!");
                0
            }
        };
        instruction /= 10;
        let out_pos: usize = match instruction % 10 {
            0 => *self.ram.entry(self.instruction_pointer + 3).or_insert(0) as usize,
            1 => self.instruction_pointer + 3,
            2 => (*self.ram.entry(self.instruction_pointer + 3).or_insert(0) + self.relative_base as i64) as usize,
            _ => {
                println!("Invalid output mode!");
                0
            }
        };
        match opcode {
            99 => self.halt(),
            1 => self.add(in1_pos, in2_pos, out_pos),
            2 => self.multiply(in1_pos, in2_pos, out_pos),
            3 => self.input(in1_pos),
            4 => self.output(in1_pos),
            5 => self.jump_if_true(in1_pos, in2_pos),
            6 => self.jump_if_false(in1_pos, in2_pos),
            7 => self.less_than(in1_pos, in2_pos, out_pos),
            8 => self.equals(in1_pos, in2_pos, out_pos),
            9 => self.offset_relative_base(in1_pos),
            _ => {
                println!("Invalid opcode!");
            }
        }
    }

    fn halt(&mut self) {
        self.instruction_pointer += 1;
        self.state = State::HALT;
    }

    fn add(&mut self, in1_pos: usize, in2_pos: usize, out_pos: usize) {
        self.ram.entry(in1_pos).or_insert(0);
        self.ram.entry(in2_pos).or_insert(0);
        self.ram.insert(out_pos, *self.ram.get(&in1_pos).expect("") + *self.ram.get(&in2_pos).expect(""));
        self.instruction_pointer += 4;
    }

    fn multiply(&mut self, in1_pos: usize, in2_pos: usize, out_pos: usize) {
        self.ram.entry(in1_pos).or_insert(0);
        self.ram.entry(in2_pos).or_insert(0);
        self.ram.insert(out_pos, *self.ram.get(&in1_pos).expect("") * *self.ram.get(&in2_pos).expect(""));
        self.instruction_pointer += 4;
    }

    fn input(&mut self, in_pos: usize) {
        self.ram.insert(in_pos, self.inputs.pop().expect("Couldn't find any input!"));
        self.instruction_pointer += 2;
    }

    fn output(&mut self, in_pos: usize) {
        self.outputs.push(*self.ram.entry(in_pos).or_insert(0));
        self.instruction_pointer += 2;
        self.state = State::WAIT;
    }

    fn jump_if_true(&mut self, in1_pos: usize, in2_pos: usize) {
        if *self.ram.entry(in1_pos).or_insert(0) != 0 {
            self.instruction_pointer = *self.ram.entry(in2_pos).or_insert(0) as usize;
        }
        else {
            self.instruction_pointer += 3;
        }
    }

    fn jump_if_false(&mut self, in1_pos: usize, in2_pos: usize) {
        if *self.ram.entry(in1_pos).or_insert(0) == 0 {
            self.instruction_pointer = *self.ram.entry(in2_pos).or_insert(0) as usize;
        }
        else {
            self.instruction_pointer += 3;
        }
    }

    fn less_than(&mut self, in1_pos: usize, in2_pos: usize, out_pos: usize) {
        if *self.ram.entry(in1_pos).or_insert(0) < *self.ram.entry(in2_pos).or_insert(0) {
            self.ram.insert(out_pos, 1);
        }
        else {
            self.ram.insert(out_pos, 0);
        }
        self.instruction_pointer += 4;
    }

    fn equals(&mut self, in1_pos: usize, in2_pos: usize, out_pos: usize) {
        if *self.ram.entry(in1_pos).or_insert(0) == *self.ram.entry(in2_pos).or_insert(0) {
            self.ram.insert(out_pos, 1);
        }
        else {
            self.ram.insert(out_pos, 0);
        }
        self.instruction_pointer += 4;
    }

    fn offset_relative_base(&mut self, in_pos: usize) {
        self.relative_base = (*self.ram.entry(in_pos).or_insert(0) + self.relative_base as i64) as usize;
        self.instruction_pointer += 2;
    }
}

fn main() {
    let program_text = fs::read_to_string("../data/Day09_input.csv").expect("Couldn't read the file!");
    // let program_text = "109,1,204,-1,1001,100,1,100,1008,100,16,101,1006,101,0,99";
    let program_text = program_text.trim();
    let program_vec: Vec<i64> = program_text.split(",").map(|x| x.parse::<i64>().expect("Couldn't parse instruction!")).collect();
    let program: HashMap<usize, i64> = (0..program_vec.len()).zip(program_vec.iter()).map(|(k, v)| (k, *v)).collect();
    
    let mut my_computer = Computer {
        ram: HashMap::new(),
        state: State::IDLE,
        instruction_pointer: 0,
        relative_base: 0,
        inputs: Vec::new(),
        outputs: Vec::new(),
    };

    // Part 1
    my_computer.load_program(&program);
    my_computer.inputs.push(1);
    my_computer.initialize();
    loop {
        match my_computer.state {
            State::HALT => break,
            _ => my_computer.run(),
        }
    }
    println!("{}", my_computer.outputs.pop().expect(""));

    // Part 2
    my_computer.load_program(&program);
    my_computer.inputs.push(2);
    my_computer.initialize();
    loop {
        match my_computer.state {
            State::HALT => break,
            _ => my_computer.run(),
        }
    }
    println!("{}", my_computer.outputs.pop().expect(""));
}
